File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {

    // Directions arrays for 8 possible moves (left, right, up, down, and 4 diagonal moves).
    int[] rowDir = {1, -1, 0, 0};
    int[] colDir = {0, 0, 1, -1};

    public int maxAreaOfIsland(int[][] grid) {
        int max_area = 0;
        int n = grid.length;
        int m = grid[0].length;

        // Traverse the entire grid
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                // If we find a 1, initiate BFS to find the area of the region.
                if (grid[i][j] == 1)
                {
                    int area = bfs(grid, i, j, n, m);
                    max_area = Math.max(max_area, area); // Track the maximum area found.
                }
            }
        }
        return max_area;
    }

    // BFS function to explore the grid in 8 directions and calculate the area of the region.
    public int bfs(int[][] grid, int row, int col, int n, int m)
    {
        // Initialize a queue for BFS.
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{row, col});
        
        // Mark the starting cell as visited by setting it to 0.
        grid[row][col] = 0;
        
        // Initialize the area of the current region as 1 (starting with the current cell).
        int area = 1;
        
        // Perform BFS.
        while (!queue.isEmpty())
        {
            // Get the current cell's coordinates.
            int[] cell = queue.poll();
            int currRow = cell[0];
            int currCol = cell[1];
            
            // Explore all 8 directions from the current cell.
            for (int i = 0; i < 4; i++)
            {
                int newRow = currRow + rowDir[i];
                int newCol = currCol + colDir[i];
                
                // Check if the new position is valid and contains a 1.
                if (isValid(grid, newRow, newCol, n, m))
                {
                    // Mark the cell as visited by setting it to 0.
                    grid[newRow][newCol] = 0;
                    // Add the cell to the queue for further exploration.
                    queue.add(new int[]{newRow, newCol});
                    // Increase the area of the region.
                    area++;
                }
            }
        }
        
        return area;
    }

    // Helper function to check if a cell is within the bounds of the grid and contains a 1.
    public boolean isValid(int[][] grid, int row, int col, int n, int m)
    {
        return (row >= 0 && row < n && col >= 0 && col < m && grid[row][col] == 1);
    }
}

File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
public:
    // DFS function to calculate the area of the island
    int dfs(vector<vector<int>>& grid, int i, int j) {
        int n = grid.size();
        int m = grid[0].size();

        // Base case: If the current cell is out of bounds or is not 1, return 0
        if (i < 0 || i >= n || j < 0 || j >= m || grid[i][j] != 1) {
            return 0;
        }

        // Mark the current cell as visited (0)
        grid[i][j] = 0;

        // Initialize the area of the current island as 1 (current cell)
        int area = 1;

        // Recursively visit all four neighboring cells (up, down, left, right)
        area += dfs(grid, i - 1, j); // up
        area += dfs(grid, i + 1, j); // down
        area += dfs(grid, i, j - 1); // left
        area += dfs(grid, i, j + 1); // right

        return area;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();

        int maxArea = 0; // Initialize the maximum area

        // Traverse the entire grid
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    // Perform DFS to calculate the area of the current island
                    int currentArea = dfs(grid, i, j);
                    // Update the maximum area
                    maxArea = max(maxArea, currentArea);
                }
            }
        }

        return maxArea;
    }
};



/*
class Solution {
public:
    //use basic dfs operation on the  grid
    int dfs(vector<vector<int>>&grid,int i,int j,int n,int m)
    {
	//checking corner cases for moving in 4 direction up down left right
        if(i>=0 and i<n and j>=0 and j<m and grid[i][j]==1)
        {
            grid[i][j]=0;//updating grid value
            //counting all the island  
            return 1+ dfs(grid,i+1,j,n,m)+dfs(grid,i-1,j,n,m)+dfs(grid,i,j+1,n,m)+dfs(grid,i,j-1,n,m);
            
        }
        return 0;
    }
    
    
    
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        
        int max_area=0;
        int n=grid.size();
        int m=grid[0].size();
        //moving every point of grid  and updating its value when you travelled
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==1)
                {
                    int count=dfs(grid,i,j,n,m);
					//finding maximum area
                    max_area=max(max_area,count);
                }
            }
        }
        return max_area;
    }
};

*/